move dmd.cond.StaticForeach.prepare to dmd.statementsem

Emmankoko opened the pull request:
`prepare ` function in the staticForeach class node requires semantic information(on expressions) and so essential to be moved to `statementsem` since prepare is called there and a UFCS advantage still being able to be called/access through a `StaticForeach` object.

Commit:
move dmd.cond.StaticForeach.prepare to dmd.statementsem

thewilsonator commented on compiler/src/dmd/statementsem.d:
```diff
@@ -5039,3 +5039,38 @@ bool checkLabel(GotoStatement gs)
+            fs.lowerNonArrayAggregate(sc);
+        }
+    }
+}
```
please add a newline at EOF

Emmankoko replied:
Done!

thewilsonator reviewed (approved):
Otherwise looks good.

RazvanN7 reviewed (changes requested):
`lowerArrayAggregate` and `lowerNonArrayAggregate` are called solely from `prepare`. You should also move them along with `prepare` to `statementsem.d`. That way they can remain private.

RazvanN7 commented on compiler/src/dmd/statementsem.d:
```diff
@@ -5039,3 +5039,38 @@ bool checkLabel(GotoStatement gs)
+* to finally expand the `static foreach` using
+* `dmd.statementsem.makeTupleForeach`.
+*/
+extern(D) void prepare(StaticForeach fs, Scope* sc)
```
The extern(D) is implicit here. When `prepare` was a member function this was necessary because the class that contained is extern(C++), but it's not necessary anymore.

Commit:
move lowerNonArrayAggregate and lowerArrayAggregate from dmd.cond.StaticForeach to dmd.statementsem

since they are both called solely from prepare which is in dmd.statementsem now.
and we can get them both as private functions.
lowerNonArrayAggregate is also an extern(C++) function which has to be explicitly declared in statementsem.

Emmankoko commented:
> `lowerArrayAggregate` and `lowerNonArrayAggregate` are called solely from `prepare`. You should also move them along with `preapre` to `statementsem.d`. That way they can remain private.

check this commit out. I don't know where the segfaults is coming from but maybe you can give a better view.

https://github.com/dlang/dmd/pull/16265/commits/f00aad6a5929ed5fceca1f485fb2b8fd3ae3a254

Commit:
move CreateForeach, wrapAndcall, createTuple, and createTypeTuple to statementsem.d

these four functions are called solely in lowerArrayAggregate and lowerNonArrayAggregate so
bringing them statementsem still keeps them private.
since they are no more member functions of an extern(C++) class, extern(D) declaration becomes unecessary
